### PR TITLE
A simplified KVUtils ledger, implemented on SQL with JDBC.

### PR DESCRIPTION
**Pull request by SamirTalwar**
This is the first draft of an SQL-based ledger. The end goal is to replace the current Sandbox implementations with something stemming from this.

Right now it's tested with [H2](https://h2database.com/) and [SQLite](https://sqlite.org/).

The unit tests and conformance tests pass on SQLite. On H2, there are a few conformance tests that are timing out… not really sure why yet.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        val (logEntry, stateUpdates) = KeyValueCommitting.processSubmission(
+          engine,
+          entryId,
+          currentRecordTime(),
+          LedgerReader.DefaultTimeModel,
+          submission,
+          participantId,
+          stateInputs,
+        )
+        verifyStateUpdatesAgainstPreDeclaredOutputs(stateUpdates, entryId, submission)
```
Comment (dajmaki) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: This assertion should move into `processSubmission` here and removed from here and from the other places that have this copied (file, in-memory).

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      sys.error(
+        s"CommitActor: State updates not a subset of expected updates! Keys [$unaccountedKeys] are unaccounted for!")
+    }
+  }
+
+  private def currentRecordTime(): Timestamp =
+    Timestamp.assertFromInstant(Clock.systemUTC().instant())
+
+  private def allocateEntryId(): DamlLogEntryId = {
+    val nonce: Array[Byte] = Array.ofDim(8)
```
Comment (dajmaki) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: kvutils should probably provide a canonical way to allocate a random entry id with enough entropy. This way of allocating doesn't provide enough entropy. For the in-memory and SQL implementations we might want to consider just allocating the ids sequentially.

kvutils also doesn't currently declare the outputs as inputs and hence does not check for conflicts on outputs itself (e.g. we might end up overwriting an existing log entry!). We might want to consider doing this check in kvutils itself. FYI @miklos-da.

Comment (dajmaki) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: It does have the downside that we end up doing more fetches for keys that we expect not to exist and depending on the ledger we might have other options for detecting the conflict (e.g. allocating entry ids in a conflict free way).

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala (hunk 1 above)
Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Agreed, just wanted to discuss this with you and @miklos-da first.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala (hunk 2 above)
Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: You're right about just using sequential IDs here. Will fix at some point.

Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Regarding the checks, I think it's a good idea. Perhaps the implementation can decide whether to add it to the list of inputs or not.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        offset.getOrElse(FirstOffset).components.head,
+        RangeSource((start, end) =>
+          withConnection { implicit connection =>
+            Future(Source(queries.selectFromLog(start, end)))
+        })
+      )
+      .map { case (_, record) => record }
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    withConnection { implicit connection =>
```
Comment (miklos-da) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: The operations below (read / update state) should be performed in a single transaction.

Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: They are. We `setAutoCommit(false)` below, which means everything happens in a transaction. Acquiring the connection starts the transaction, and closing it implicitly commits. I'm not a big fan of implicit behavior, so [`withConnection`](https://github.com/digital-asset/daml/pull/3940/files#diff-9f45aadbe621258428069b8880a87344R160) explicitly commits if all goes well. It also catches exceptions and explicitly rolls back.

Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Let's rename this to `inTransaction`.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/Main.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          (jdbcUrl, config) => config.copy(extra = config.extra.copy(jdbcUrl = Some(jdbcUrl))))
+      ()
+    }
+
+    override def apply(participantId: ParticipantId, config: ExtraConfig)(
+        implicit materializer: Materializer,
+    ): KeyValueLedger =
+      Await.result(
+        SqlLedgerReaderWriter(participantId = participantId, jdbcUrl = config.jdbcUrl.getOrElse {
+          throw new IllegalStateException("No root directory provided.")
```
Comment (gerolf-da) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/Main.scala]:
```suggestion
          throw new IllegalStateException("No jdbc URL provided.")
```

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala (hunk 3 above)
Comment (gerolf-da) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: I suggest adding `DB` or `Database` (e.g. `inDbTransaction`) to the name as well. There are enough different kinds of transactions that warrant making it super clear which one is meant.

Comment (miklos-da) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Thanks for the clarification @SamirTalwar.

File: ledger/ledger-on-sql/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        ],
+    },
+    {
+        "name": "sqlite",
+        "runtime_dep": "@maven//:org_xerial_sqlite_jdbc",
+        "conformance_test_server_main": "com.daml.ledger.on.sql.MainWithEphemeralSqlite",
+    },
+]
+
+compile_deps = [
```
[the remark this answers is not in the document]
Comment (SamirTalwar) [reply ledger/ledger-on-sql/BUILD.bazel]: Done as part of rebasing and verifying that every commit is valid.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala (hunk 3 above)
Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: Renamed it to `inDatabaseTransaction`, and renamed the other `withConnection` to `withDatabaseStream`.

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/LedgerConstructor.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.app
+
+import akka.stream.Materializer
+import com.daml.ledger.participant.state.v1.ParticipantId
+import scopt.OptionParser
+
+trait LedgerConstructor[ExtraConfig] {
```
[the remark this answers is not in the document]
Comment (SamirTalwar) [reply ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/LedgerConstructor.scala]: Renamed it. 😄

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala (hunk 1 above)
Comment (miklos-da) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala]: I have added a ticket to track this work in the FCL Jira.

**Comment by SamirTalwar** [comment]
I have disabled the H2 conformance tests for now until we can figure out why some of them time out.

**Comment by SamirTalwar** [comment]
Currently, we cap SQLite to a single connection; the connection pool is only used to enforce this across threads, because `getConnection` will block if there are no more connections left in the pool.

There seems to be a way to use `SQLiteConnectionPoolDataSource` to have multiple connections using the same underlying physical connection, but I can't see how it works yet.

**Pull request merged**